### DISCUSSION TITLE
HasOne Linq select query issue

### DISCUSSION THREAD

**Discussion by BrennanVanCoillie** [nhibernate/nhibernate-core#3287] [Q&A] [answered]
### Hi, We have an issue when using a one to one(or zero) relationship in our code

For example we have a class Parent and class Child which have one to one or zero relation

This the Parent class : 
```
public class Parent
{
    public virtual Guid Id { get; set; }
    public virtual Child Child { get; set; }
}
```
This is the Child class
```
public class Child
{
    public virtual Guid Id { get; set; }
    public virtual Parent Parent { get; set; }
}
```
This is the mapping we use for the parent class
```
public class ParentMapping : ClassMap<Parent>
{
    public ParentMapping()
    {
        Id(x => x.Id);
        HasOne(x => x.Child)
            .PropertyRef(x => x.Parent)
            .Cascade.All()
            .LazyLoad();
    }
}
```
This is the mapping we use for the child class
```
public class ChildMapping : ClassMap<Child>
{
    public ChildMapping()
    {
        Id(x => x.Id);
        References(x => x.Parent)
            .Unique()
            .Not.Nullable();
    }
}
```

Then we try getting all the parent records with their child records.

The problem is that when we query the database using `.Query<Parent>().ToList()` Nhibernate generates a query for the `Parent` and then a single query for every `Child` record that is found.
```
public IEnumerable<Parent> TestMethodUsingQuery()
    {
        return Session.Query<Parent>().ToList();
    }
```

When we use `QueryOver<>().List()` however Nhibernate generates a `left outer join` to fetch all the data, which for us is as expected.
```
    public IEnumerable<Parent> TestMethodUsingQueryOver()
    {
        return Session.QueryOver<Parent>().List();
    }
```
Can someone please help us on this issue?
This was tested using `Nhibernate 5.4.2` and `FLuentNHibernate 3.1.0`

**Comment by gliljas** [comment, marked as answer]
`return Session.Query<Parent>().Fetch(x=>x.Child).ToList();`

**Reply by BrennanVanCoillie** [reply]
This indeed gives the expected result.
Could this be solved with mapping rather then putting fetch in the query?

If not...
We have quite a lot of issues when trying to test this when just using it as an extension in a query on a list.
We get the following error.

```
System.InvalidOperationException : There is no method 'Fetch' on type 'NHibernate.Linq.EagerFetchingExtensionMethods' that matches the specified arguments
   at System.Linq.EnumerableRewriter.FindMethod(Type type, String name, ReadOnlyCollection`1 args, Type[] typeArgs)
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression m)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression m)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
   at System.Linq.EnumerableRewriter.VisitMethodCall(MethodCallExpression m)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at System.Linq.EnumerableQuery`1.GetEnumerator()
   at System.Linq.EnumerableQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at
 ```
   
Example in how we query it : 

```
IEnumerable<Parent> parents = new List<Parent>().AsQueryable();
        parents
            .Fetch(x => x.Child)
            .Where(x => x.Amount > 0)
            .OrderByDescending(x => x.Date)
            .ToList();
```
We are just testing what the query returns here, there is no session used!
Is there any way to make this work, or is this not possible?
Thanks a lot!

**Reply by Madajevas** [reply]
Maybe this will help https://github.com/nhibernate/nhibernate-core/issues/3292

**Comment by Madajevas** [comment]
I am having the same issue. Problem is, parent entities are queried in quite a lot of locations, project is maintained by a team of developers and new queries may pop up any day. Also, info contained in child is needed only for a fraction of scenarios.

I have tried adding `.Constrained()` to has one mapping and it does work (create proxies) if no `.PropertyRef` is there. In this case id fields of both entities must match and with amount of data we have it would be preferable not to update primary key and have a join of "parent_id" field.

Can this be achieved somehow with mapping? My effort to google this led nowhere.

UPDATE:
`.Constrained()` with `.PropertyRef` seems to working when used with `QueryOver` API. The very same mapping with `Query` produces N+1.
UPDATE#2:
Assumption that `QueryOver` acts differently was false.
